File: bot/src/main/java/by/yayauheny/tutochkatgbot/dto/backend/SubwayStationResponseDto.java
package by.yayauheny.tutochkatgbot.dto.backend;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.time.Instant;
import java.util.UUID;

/**
 * Subway station response DTO
 */
public record SubwayStationResponseDto(
    @JsonProperty("id") UUID id,
    @JsonProperty("subwayLineId") UUID subwayLineId,
    @JsonProperty("nameRu") String nameRu,
    @JsonProperty("nameEn") String nameEn,
    @JsonProperty("isTransfer") Boolean isTransfer,
    @JsonProperty("coordinates") LatLon coordinates,
    @JsonProperty("isDeleted") Boolean isDeleted,
    @JsonProperty("createdAt") Instant createdAt,
    @JsonProperty("line") SubwayLineResponseDto line
) {
    public String displayName(String preferredLang) {
        if ("en".equalsIgnoreCase(preferredLang) && nameEn != null && !nameEn.isBlank()) {
            return nameEn.trim();
        }
        return nameRu != null && !nameRu.isBlank() ? nameRu.trim() : (nameEn != null ? nameEn.trim() : "Unknown");
    }

    public String lineColor() {
        return line != null && line.hexColor() != null && !line.hexColor().isBlank()
            ? line.hexColor()
            : null;
    }
}

File: common/src/main/kotlin/yayauheny/by/model/subway/SubwayStationResponseDto.kt
package yayauheny.by.model.subway

import java.time.Instant
import java.util.UUID
import kotlin.jvm.JvmOverloads
import kotlinx.serialization.Contextual
import kotlinx.serialization.Serializable
import yayauheny.by.model.dto.Coordinates

@Serializable
data class SubwayStationResponseDto(
    @Contextual val id: UUID,
    @Contextual val subwayLineId: UUID?,
    val nameRu: String,
    val nameEn: String,
    val isTransfer: Boolean = false,
    val coordinates: Coordinates,
    val isDeleted: Boolean,
    @Contextual val createdAt: Instant,
    val line: SubwayLineResponseDto? = null
) {
    @JvmOverloads
    fun displayName(preferredLang: String? = null): String {
        return when {
            preferredLang?.equals("en", ignoreCase = true) == true && !nameEn.isBlank() -> nameEn.trim()
            !nameRu.isBlank() -> nameRu.trim()
            !nameEn.isBlank() -> nameEn.trim()
            else -> "Unknown"
        }
    }

    fun lineColor(): String? = line?.hexColor?.takeIf { it.isNotBlank() }
}
